chain/trie/pruning.go

```diff
@@ -0,0 +1,305 @@
+}
+
+func (ref *RefCounts) DumpInfos(offset uint32, pruning bool) error {
+	if !pruning {
+		fmt.Println("refcount:", time.Now().Format("2006-01-02 15:04:05"), ref.startRefCountHeight+offset, ref.dbCounter, ref.nodeCounter, len(ref.counts))
+	} else {
+		fmt.Println("pruning: ", time.Now().Format("2006-01-02 15:04:05"), ref.startPruningHeight+offset, ref.dbCounter, ref.nodeCounter, len(ref.counts))
+	}
+
+	return nil
+}
+
+func (ref *RefCounts) CreateRefCounts(hash common.Uint256) error {
+	root, err := ref.trie.resolveHash(hash.ToArray(), true)
+	if err != nil {
+		return err
+	}
+
+	ref.trie.root = root
+	ref.dbCounter = 0
+	ref.nodeCounter = 0
+	err = ref.createRefCounts(ref.trie.root)
+	if err != nil {
+		panic(err)
+	}
+
+	return nil
+}
+
+func (ref *RefCounts) createRefCounts(n node) error {
+	ref.nodeCounter++
+	switch n := n.(type) {
+	case *shortNode:
+		hash, _ := n.cache()
+		hs, _ := common.Uint256ParseFromBytes(hash)
+		_, ok := ref.counts[hs]
+		if !ok {
+			if err := ref.createRefCounts(n.Val); err != nil {
+				return err
+			}
+		}
+
+		ref.counts[hs] = ref.counts[hs] + 1
+		return nil
+	case *fullNode:
+		hash, _ := n.cache()
+		hs, _ := common.Uint256ParseFromBytes(hash)
+		_, ok := ref.counts[hs]
+		if !ok {
+			for i := 0; i < 17; i++ {
+				if n.Children[i] != nil {
+					err := ref.createRefCounts(n.Children[i])
+					if err != nil {
+						return err
+					}
+				}
+			}
+		}
+
+		ref.counts[hs] = ref.counts[hs] + 1
+		return nil
+	case hashNode:
+		hs, _ := common.Uint256ParseFromBytes(n)
+		count, ok := ref.counts[hs]
+		if ok {
+			ref.counts[hs] = count + 1
+			return nil
+		}
+		child, err := ref.trie.resolveHash(n, true)
+		if err != nil {
+			return err
+		}
+		ref.dbCounter++
+		return ref.createRefCounts(child)
+	case nil:
+		return nil
+	case valueNode:
+		return nil
+	default:
+		panic(fmt.Sprintf("invalid node type : %v, %v", reflect.TypeOf(n), n))
+
+	}
+}
+
+func (ref *RefCounts) Pruning(hash common.Uint256) error {
+	root, err := ref.trie.resolveHash(hash.ToArray(), true)
+	if err != nil {
+		return err
+	}
+	ref.trie.root = root
+
+	err = ref.pruning(ref.trie.root)
+	if err != nil {
+		panic(err)
+	}
+
+	return nil
+}
+
+func (ref *RefCounts) pruning(n node) error {
+	switch n := n.(type) {
+	case *shortNode:
+		hash, _ := n.cache()
+		hs, _ := common.Uint256ParseFromBytes(hash)
+
+		count, ok := ref.counts[hs]
+		if !ok {
+			panic("cannot get refcount")
+		}
+		if count == 0 {
+			panic(fmt.Sprintf("refCount cannot be zero1, %v", hs.ToHexString()))
+		}
+
+		if count > 1 {
+			ref.counts[hs] = count - 1
+			return nil
+		}
+
+		delete(ref.counts, hs)
+		ref.trie.db.BatchDelete(db.TrieNodeKey([]byte(hash)))
+		return ref.pruning(n.Val)
+
+	case *fullNode:
+		hash, _ := n.cache()
+		hs, _ := common.Uint256ParseFromBytes(hash)
+
+		count, ok := ref.counts[hs]
+		if !ok {
+			panic("cannot get refcount")
+		}
+		if count == 0 {
+			panic(fmt.Sprintf("refCount cannot be zero2, %v", hs.ToHexString()))
```

Comment:
"zero2" -> "zero"
We can use "shortNode refCount should not be zero" and "fullNode refCount should not be zero" to distinguish these two